libyul/optimiser/RedundantStoreEliminator.cpp

```diff
@@ -0,0 +1,436 @@
+				}
+			};
+		}
+		case Instruction::CREATE:
+		case Instruction::CREATE2:
+			return vector<Operation>{
+				Operation{
+					Location::Memory,
+					Effect::Read,
+					identifierNameIfSSA(_functionCall.arguments.at(1)),
+					identifierNameIfSSA(_functionCall.arguments.at(2))
+				},
+				// Unknown read includes unknown write.
+				Operation{Location::Storage, Effect::Read, {}, {}},
+			};
+		case Instruction::MSIZE:
+			// This is just to satisfy the assert below.
+			return vector<Operation>{};
+		default:
+			yulAssert(
+				evmasm::SemanticInformation::storage(*instruction) ==
+				evmasm::SemanticInformation::None,
+				""
+			);
+			yulAssert(
+				evmasm::SemanticInformation::memory(*instruction) ==
+				evmasm::SemanticInformation::None,
+				""
+			);
+			yulAssert(
+				sideEffects.memory == SideEffects::Effect::None &&
+				sideEffects.storage == SideEffects::Effect::None,
+				""
+			);
+
+			break;
+		}
+	}
+
+	vector<Operation> result;
+	if (sideEffects.memory != SideEffects::Effect::None)
+		result.emplace_back(Operation{Location::Memory, Effect::Read, {}, {}});
+	if (sideEffects.storage != SideEffects::Effect::None)
+		result.emplace_back(Operation{Location::Storage, Effect::Read, {}, {}});
+
+	return result;
+}
+
+void RedundantStoreEliminator::applyOperation(RedundantStoreEliminator::Operation const& _operation)
+{
+	for (auto& [statement, state]: m_stores[YulString{}])
+		if (state == State::Undecided)
+		{
+			Operation const& storeOperation = m_storeOperations.at(statement);
+			if (_operation.effect == Effect::Read && !knownUnrelated(storeOperation, _operation))
+				state = State::Used;
+			else if (_operation.effect == Effect::Write && knownCovered(storeOperation, _operation))
+				state = State::Unused;
+		}
+}
+
+bool RedundantStoreEliminator::knownUnrelated(
+	RedundantStoreEliminator::Operation const& _op1,
+	RedundantStoreEliminator::Operation const& _op2
+) const
+{
+	KnowledgeBase knowledge(m_dialect, m_ssaValues);
+
+	if (_op1.location != _op2.location)
+		return true;
+	if (_op1.location == Location::Storage)
+	{
+		if (_op1.start && _op2.start)
+			return knowledge.knownToBeDifferent(*_op1.start, *_op2.start);
+	}
+	else
+	{
+		u256 largestPositive = (u256(1) << 128) - 1;
+
+		yulAssert(_op1.location == Location::Memory, "");
+		if (_op1.length && knowledge.knownToBeZero(*_op1.length))
+			return true;
+		if (_op2.length && knowledge.knownToBeZero(*_op2.length))
+			return true;
+		if (!_op1.start || !_op2.start || !_op1.length || !_op2.length)
+			// Can we say anything else here?
+			return false;
+
+		// 1.start + 1.length <= 2.start ||
+		if (optional<u256> length1 = knowledge.valueIfKnownConstant(*_op1.length))
+			if (optional<u256> diff = knowledge.differenceIfKnownConstant(*_op2.start, *_op1.start))
+				// diff = 2.start - 1.start
+				if (*length1 <= *diff && *length1 <= largestPositive && *diff <= largestPositive)
+					return true;
```

Comment:
The subtraction here will overflow if `*_op1.start > *_op2.start`. When the variables are close to the opposite ends of addressable memory the result could be a number small enough to fit under `length` and cause a false-positive.

I guess it's unlikely that anyone will be able to write to memory close to the limit of `u256` but might be better to have a check against this just in case. Or at least a comment saying that we're aware of it and ignoring it on purpose.

Reply:
But isn't that what the `*diff <= largestPositive` is for?

Reply:
Oh, I didn't notice that `largestPositive` was meant to be the largest signed positive number and not just 2^256-1.

But in that case I think the check is not fully effective. I think it would only work if the difference was guaranteed to never go below -2^128. If you have one slot at 0 and the other at 2^256-32, you'll get a bigger difference (not even representable with a signed number), which will result in a very small positive number when it underflows.

Reply:
Can you make a test case out of it?